src/watchdog.rs

```diff
@@ -0,0 +1,185 @@
+#[cfg(esp_idf_esp_task_wdt)]
+pub mod task {
+    //! ## Example
+    //!
+    //! ```rust, ignore
+    //! # fn main() -> Result<()> {
+    //! let mut watchdog = TaskWatchdog::take()?.unwrap();
+    //!
+    //! watchdog.reconfigure(WatchdogConfig {
+    //!     duration: Duration::from_secs(2),
+    //!     panic_on_trigger: true,
+    //!     ..Default::default()
+    //! })?;
+    //!
+    //! let task = Task::current();
+    //!
+    //! task.set_watchdog(&mut watchdog)?;
+    //!
+    //! loop {
+    //!     task.reset_watchdog()?;
+    //!     unsafe { vTaskDelay(1) };
+    //! }
+    //! # }
+    //! ```
+
+    use esp_idf_sys::*;
+
+    use crate::cpu::Core;
+
+    pub struct WatchdogConfig {
+        pub duration: core::time::Duration,
+        pub panic_on_trigger: bool,
+        // unused by now
+        pub cores: heapless::Vec<Core, 2>,
+    }
+
+    impl Default for WatchdogConfig {
+        fn default() -> Self {
+            let mut cores = heapless::Vec::new();
+            if CONFIG_ESP_TASK_WDT_CHECK_IDLE_TASK_CPU0 == 1 {
+                cores.push(Core::Core0).unwrap();
+            }
+            #[cfg(any(esp32, esp32s3))]
+            if CONFIG_ESP_TASK_WDT_CHECK_IDLE_TASK_CPU1 == 1 {
+                cores.push(Core::Core1).unwrap();
+            }
+            Self {
+                duration: core::time::Duration::from_secs(CONFIG_ESP_TASK_WDT_TIMEOUT_S as u64),
+                panic_on_trigger: false, // CONFIG_ESP_TASK_WDT_PANIC
+                cores,
+            }
+        }
+    }
+
+    #[cfg(esp_idf_version = "master")]
+    impl From<WatchdogConfig> for esp_task_wdt_config_t {
+        fn from(value: WatchdogConfig) -> Self {
+            esp_task_wdt_config_t {
+                timeout_ms: value.duration.as_millis() as u32,
+                trigger_panic: value.panic_on_trigger,
+                idle_core_mask: value
+                    .cores
+                    .iter()
+                    .fold(0u32, |mask, core| mask & (1 << (*core as u32))),
+            }
+        }
+    }
+
+    #[cfg(feature = "riscv-ulp-hal")]
+    static mut TAKEN: bool = false;
+
+    #[cfg(not(feature = "riscv-ulp-hal"))]
+    static TAKEN: core::sync::atomic::AtomicBool = core::sync::atomic::AtomicBool::new(false);
+
+    #[cfg(not(feature = "riscv-ulp-hal"))]
+    static TAKEN_CS: crate::task::CriticalSection = crate::task::CriticalSection::new();
+
+    #[derive(Debug, Clone)]
+    pub struct TaskWatchdog;
+
+    impl TaskWatchdog {
+        #[cfg(feature = "riscv-ulp-hal")]
+        pub fn take() -> Result<Option<Self>, EspError> {
+            if unsafe { TAKEN } {
+                None
+            } else {
+                unsafe {
+                    TAKEN = true;
+                }
+                Some(Self::init())
+            }
+        }
+
+        #[cfg(not(feature = "riscv-ulp-hal"))]
+        pub fn take() -> Result<Option<Self>, EspError> {
```

Comment:
For simplicity and symmetry with the rest of the HAL code, it should be `-> Result<Self, EspError>`, where if the watchdog is already taken (initialized), you should return the most appropriate ESP error.